File: mobile/src/main/java/com/siliconlabs/bledemo/toolbars/FilterFragment.java
package com.siliconlabs.bledemo.toolbars;


import android.os.Bundle;

import androidx.fragment.app.Fragment;
import androidx.core.content.ContextCompat;
import androidx.appcompat.widget.AppCompatCheckBox;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.text.Editable;
import android.text.InputFilter;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.SeekBar;
import android.widget.TextView;

import com.siliconlabs.bledemo.other.BeaconType;
import com.siliconlabs.bledemo.R;
import com.siliconlabs.bledemo.other.SavedSearch;
import com.siliconlabs.bledemo.adapters.BeaconTypeAdapter;
import com.siliconlabs.bledemo.adapters.SavedSearchesAdapter;
import com.siliconlabs.bledemo.beaconutils.BleFormat;
import com.siliconlabs.bledemo.utils.FilterDeviceParams;
import com.siliconlabs.bledemo.utils.SharedPrefUtils;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Map;


/**
 * A simple {@link Fragment} subclass.
 */
public class FilterFragment extends Fragment {

    private SeekBar rssiSeekBar;
    private TextView rssiValueText;
    private ImageView beaconArrowIV;
    private ImageView savedArrowIV;
    private ImageView closeBtn;
    private Button resetBtn;
    private Button saveBtn;
    private Button searchBtn;
    private AppCompatCheckBox onlyFavourites;
    private AppCompatCheckBox onlyConnectable;
    private RecyclerView beaconTypeRV;
    private RecyclerView savedSearchesRV;
    private BeaconTypeAdapter beaconTypeadapter;
    private SavedSearchesAdapter savedSearchesAdapter;
    private SharedPrefUtils sharedPrefUtils;
    private boolean rssiFlag;

    // UI elements for search boxes
    private EditText deviceNameET;
    private ImageView clearPacketNameSearchIV;
    private EditText packetContentET;
    private ImageView clearPacketContentSearchIV;

    public FilterFragment() {
        // Required empty public constructor
    }

    private ToolbarCallback toolbarCallback;

    public FilterFragment setCallback(ToolbarCallback toolbarCallback) {
        this.toolbarCallback = toolbarCallback;
        return this;
    }

    private void init(View view) {
        closeBtn = view.findViewById(R.id.imageview_close);
        saveBtn = view.findViewById(R.id.button_save);
        searchBtn = view.findViewById(R.id.button_search);
        resetBtn = view.findViewById(R.id.button_reset);
        rssiSeekBar = view.findViewById(R.id.seekbar_rssi);
        rssiValueText = view.findViewById(R.id.textview_rssi_value);
        beaconArrowIV = view.findViewById(R.id.imageview_beacon_arrow);
        savedArrowIV = view.findViewById(R.id.imageview_save_arrow);
        onlyFavourites = view.findViewById(R.id.only_favourites);
        onlyConnectable = view.findViewById(R.id.only_connectable);
        beaconTypeRV = view.findViewById(R.id.recyclerview_beacon_type);
        savedSearchesRV = view.findViewById(R.id.recyclerview_saved_searches);

        deviceNameET = view.findViewById(R.id.search_device_name_edit_text);
        packetContentET = view.findViewById(R.id.search_packet_content_edit_text);
        clearPacketNameSearchIV = view.findViewById(R.id.search_clear_packet_name);
        clearPacketContentSearchIV = view.findViewById(R.id.search_clear_packet_content);

        setSearchBox(deviceNameET, clearPacketNameSearchIV);
        setSearchBox(packetContentET, clearPacketContentSearchIV);
        packetContentET.setFilters(new InputFilter[]{new InputFilter.AllCaps()});

        ArrayList<BeaconType> list = new ArrayList<>();
        final List<BleFormat> bleFormats = new ArrayList<>(Arrays.asList(BleFormat.values()));
        for (BleFormat bf : bleFormats) {
            list.add(new BeaconType(getResources().getString(bf.getNameResId()), false, bf));
        }
        beaconTypeadapter = new BeaconTypeAdapter(list, getActivity());
        LinearLayoutManager beaconLinearLayoutManager = new LinearLayoutManager(getActivity());
        beaconTypeRV.setLayoutManager(beaconLinearLayoutManager);
        beaconTypeRV.addItemDecoration(new DividerItemDecoration(beaconTypeRV.getContext(), beaconLinearLayoutManager.getOrientation()));
        beaconTypeRV.setAdapter(beaconTypeadapter);

        ArrayList<SavedSearch> searchlist = new ArrayList<>();
        for (Map.Entry mapElement : sharedPrefUtils.getMapFilter().entrySet()) {
            searchlist.add(new SavedSearch((String) mapElement.getKey()));
        }
        savedSearchesAdapter = new SavedSearchesAdapter(searchlist, getContext(), new SavedSearchesAdapter.SavedSearchesCallback() {
            @Override
            public void onClick(String name) {
                FilterDeviceParams f = sharedPrefUtils.getMapFilter().get(name);
                sharedPrefUtils.setLastFilter(f);
                fillFields(f);
                savedSearchesAdapter.notifyDataSetChanged();

            }
        });
        LinearLayoutManager searchLinearLayoutManager = new LinearLayoutManager(getActivity());
        savedSearchesRV.setLayoutManager(searchLinearLayoutManager);
        savedSearchesRV.addItemDecoration(new DividerItemDecoration(savedSearchesRV.getContext(), searchLinearLayoutManager.getOrientation()));
        savedSearchesRV.setAdapter(savedSearchesAdapter);
    }


    private void setSearchBox(final EditText searchEditText, final ImageView clearImageView) {

        searchEditText.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                if (count <= 0) clearImageView.setVisibility(View.GONE);
                else clearImageView.setVisibility(View.VISIBLE);
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });

        clearImageView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                searchEditText.getText().clear();
            }
        });

    }

    private void ui() {
        closeBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                toolbarCallback.close();
            }
        });
        resetBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                resetUi();
                submit(true);
            }
        });
        saveBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String key = prepareFilterName();
                if (key.equals("")) return;
                if (!sharedPrefUtils.getMapFilter().containsKey(key)) {
                    FilterDeviceParams filterDeviceParams = prepareFilterDeviceParam();
                    sharedPrefUtils.addToMapFilterAndSave(key, filterDeviceParams);
                    sharedPrefUtils.setLastFilter(filterDeviceParams);
                    savedSearchesAdapter.addItem(new SavedSearch(key));
                    savedSearchesAdapter.notifyDataSetChanged();
                }
            }
        });
        rssiSeekBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                rssiFlag = true;
                int rssi = getRssiValue(progress);
                rssiValueText.setText(getResources().getString(R.string.n_dBm, rssi));

            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {
            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {

            }
        });
        searchBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                submit(true);
            }
        });
        beaconArrowIV.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                //If after clicking arrow is up then...
                if (beaconArrowIV.getTag().equals("ARROW_UP")) {
                    beaconArrowIV.setImageDrawable(ContextCompat.getDrawable(getActivity(), R.drawable.ic_arrow_down_on));
                    beaconTypeRV.setVisibility(View.GONE);
                    beaconArrowIV.setTag("ARROW_DOWN");
                } else {
                    beaconArrowIV.setImageDrawable(ContextCompat.getDrawable(getActivity(), R.drawable.ic_arrow_up_on));
                    beaconTypeRV.setVisibility(View.VISIBLE);
                    beaconArrowIV.setTag("ARROW_UP");
                }
            }
        });
        savedArrowIV.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                //If after clicking arrow is up then...
                if (savedArrowIV.getTag().equals("ARROW_UP")) {
                    savedArrowIV.setImageDrawable(ContextCompat.getDrawable(getActivity(), R.drawable.ic_arrow_down_on));
                    savedSearchesRV.setVisibility(View.GONE);
                    savedArrowIV.setTag("ARROW_DOWN");
                } else {
                    savedArrowIV.setImageDrawable(ContextCompat.getDrawable(getActivity(), R.drawable.ic_arrow_up_on));
                    savedSearchesRV.setVisibility(View.VISIBLE);
                    savedArrowIV.setTag("ARROW_UP");
                }
            }
        });
    }

    private String prepareFilterName() {
        StringBuilder name = new StringBuilder();
        String seperator = " + ";
        if (rssiFlag) {
            name.append("RSSI > ").append(rssiValueText.getText()).append(seperator);
        }
        List<BleFormat> selectedBeacons = getSelectedBeacons();
        if (!selectedBeacons.isEmpty()) {
            name.append("Beacon type ");
            for (int i = 0; i < selectedBeacons.size(); i++) {
                name.append("\"").append(getResources().getString(selectedBeacons.get(i).getNameResId())).append("\"");
                if (i != selectedBeacons.size() - 1) {
                    name.append(", ");
                }
            }
            name.append(seperator);
        }
        if (!deviceNameET.getText().toString().equals("")) {
            name.append("Name \"").append(deviceNameET.getText().toString()).append("\"").append(seperator);
        }
        if (!packetContentET.getText().toString().equals("")) {
            name.append("Packet content \"").append(packetContentET.getText().toString()).append("\"").append(seperator);
        }
        if (onlyFavourites.isChecked()) {
            name.append("Only favourites").append(seperator);
        }
        if (onlyConnectable.isChecked()) {
            name.append("Only connectable");
        }
        if (name.toString().endsWith(seperator)) {
            name = new StringBuilder(name.substring(0, name.length() - 3));
        }
        return name.toString();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_filter, container, false);
        sharedPrefUtils = new SharedPrefUtils(getContext());
        init(view);
        ui();

        if (sharedPrefUtils.getLastFilter() != null) {
            fillFields(sharedPrefUtils.getLastFilter());
        } else {
            resetUi();
        }

        return view;
    }

    private int getRssiValue(int progress) {
        return progress - 100;
    }

    private void resetUi() {
        deviceNameET.setText("");
        packetContentET.setText("");
        rssiSeekBar.setProgress(0);
        rssiValueText.setText(R.string.dialog_device_filter_rssi_not_set);
        rssiFlag = false;
        onlyFavourites.setChecked(false);
        onlyConnectable.setChecked(false);
        beaconTypeadapter.selectBeacons(Collections.<BleFormat>emptyList());
    }

    private void fillFields(FilterDeviceParams filterDeviceParams) {
        deviceNameET.setText(filterDeviceParams.getName());
        packetContentET.setText(filterDeviceParams.getAdvertising());
        rssiSeekBar.setProgress(filterDeviceParams.getRssiValue() + 100);
        rssiSeekBar.refreshDrawableState();
        int rssi = filterDeviceParams.getRssiValue();
        rssiFlag = filterDeviceParams.isRssiFlag();
        rssiValueText.setText(rssiFlag ? getResources().getString(R.string.n_dBm, rssi) : getContext().getString(R.string.dialog_device_filter_rssi_not_set));
        List<BleFormat> selectedBleFormats = filterDeviceParams.getBleFormats();
        beaconTypeadapter.selectBeacons(selectedBleFormats);
        onlyFavourites.setChecked(filterDeviceParams.isOnlyFavourite());
        onlyConnectable.setChecked(filterDeviceParams.isOnlyConnectable());

    }

    private List<BleFormat> getSelectedBeacons() {
        List<BleFormat> selectedBeacons = new ArrayList<>();
        for (BeaconType b : beaconTypeadapter.getBeaconTypeList()) {
            if (b.isChecked()) selectedBeacons.add(b.getBleFormat());
        }
        return selectedBeacons;
    }

    private FilterDeviceParams prepareFilterDeviceParam() {
        String name = deviceNameET.getText().toString();
        String advertising = packetContentET.getText().toString();
        int rssi = getRssiValue(rssiSeekBar.getProgress());
        return new FilterDeviceParams(prepareFilterName(), name, advertising, rssi, rssiFlag, getSelectedBeacons(), onlyFavourites.isChecked(), onlyConnectable.isChecked());
    }

    private void submit(boolean closeToolbar) {
        FilterDeviceParams filterDeviceParams = prepareFilterDeviceParam();
        sharedPrefUtils.setLastFilter(filterDeviceParams);
        toolbarCallback.submit(filterDeviceParams, closeToolbar);
    }

}

File: mobile/src/main/java/com/siliconlabs/bledemo/Browser/Fragments/FilterFragment.kt
package com.siliconlabs.bledemo.Browser.Fragments

import android.os.Bundle
import android.text.Editable
import android.text.InputFilter
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import android.widget.ImageView
import android.widget.SeekBar
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import com.siliconlabs.bledemo.BeaconUtils.BleFormat
import com.siliconlabs.bledemo.Browser.Adapters.BeaconTypeAdapter
import com.siliconlabs.bledemo.Browser.Adapters.SavedSearchesAdapter
import com.siliconlabs.bledemo.Browser.Adapters.SavedSearchesAdapter.SavedSearchesCallback
import com.siliconlabs.bledemo.Browser.Models.BeaconType
import com.siliconlabs.bledemo.Browser.Models.SavedSearch
import com.siliconlabs.bledemo.Browser.ToolbarCallback
import com.siliconlabs.bledemo.R
import com.siliconlabs.bledemo.Utils.FilterDeviceParams
import com.siliconlabs.bledemo.Utils.SharedPrefUtils
import kotlinx.android.synthetic.main.fragment_filter.*
import java.util.*

class FilterFragment : Fragment() {
    private var toolbarCallback: ToolbarCallback? = null
    private lateinit var beaconTypeadapter: BeaconTypeAdapter
    private lateinit var savedSearchesAdapter: SavedSearchesAdapter
    private lateinit var sharedPrefUtils: SharedPrefUtils
    private var rssiFlag = false

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_filter, container, false)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        sharedPrefUtils = SharedPrefUtils(requireContext())
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        setSearchBox(et_search_device_name, iv_clear_device_name)
        setSearchBox(et_search_packet_content, iv_clear_packet_content)
        et_search_packet_content.filters = arrayOf<InputFilter>(InputFilter.AllCaps())

        initBeaconRecyclerView()
        initSavedSearchesRecyclerView()

        setListeners()
        resetUi()
    }

    private fun initBeaconRecyclerView() {
        val list = ArrayList<BeaconType>()
        val bleFormats: List<BleFormat> = ArrayList(listOf(*BleFormat.values()))
        for (bf in bleFormats) {
            list.add(BeaconType(resources.getString(bf.nameResId), false, bf))
        }

        beaconTypeadapter = BeaconTypeAdapter(list, requireContext())
        rv_beacon_type.layoutManager = LinearLayoutManager(activity)
        rv_beacon_type.addItemDecoration(DividerItemDecoration(rv_beacon_type.context, LinearLayoutManager.VERTICAL))
        rv_beacon_type.adapter = beaconTypeadapter
    }

    private fun initSavedSearchesRecyclerView() {
        val searchList = ArrayList<SavedSearch>()
        for ((key) in sharedPrefUtils.mapFilter) {
            searchList.add(SavedSearch((key as String)))
        }
        savedSearchesAdapter = SavedSearchesAdapter(searchList, requireContext(), object : SavedSearchesCallback {
            override fun onClick(name: String?) {
                val f = sharedPrefUtils.mapFilter[name]
                sharedPrefUtils.lastFilter = f
                f?.let { fillFields(f) }
                savedSearchesAdapter.notifyDataSetChanged()
            }

        })

        rv_saved_searches.layoutManager = LinearLayoutManager(activity)
        rv_saved_searches.addItemDecoration(DividerItemDecoration(rv_saved_searches.context, LinearLayoutManager.VERTICAL))
        rv_saved_searches.adapter = savedSearchesAdapter
    }


    fun setCallback(toolbarCallback: ToolbarCallback?): FilterFragment {
        this.toolbarCallback = toolbarCallback
        return this
    }

    private fun setSearchBox(etSearch: EditText, ivClear: ImageView) {
        etSearch.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                ivClear.visibility = if (count <= 0) View.GONE else View.VISIBLE
            }

            override fun afterTextChanged(s: Editable) {}
        })
        ivClear.setOnClickListener { etSearch.text.clear() }
    }

    private fun setListeners() {
        iv_close.setOnClickListener { toolbarCallback?.close() }
        btn_reset.setOnClickListener { resetUi(); submit(true) }
        btn_save.setOnClickListener(View.OnClickListener {
            val key = prepareFilterName()
            if (key == "") return@OnClickListener
            if (!sharedPrefUtils.mapFilter.containsKey(key)) {
                val filterDeviceParams = prepareFilterDeviceParam()
                sharedPrefUtils.addToMapFilterAndSave(key, filterDeviceParams)
                sharedPrefUtils.lastFilter = filterDeviceParams
                savedSearchesAdapter.addItem(SavedSearch(key))
                savedSearchesAdapter.notifyDataSetChanged()
            }
        })

        sb_rssi.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                rssiFlag = true
                val rssi = getRssiValue(progress)
                tv_rssi_value.text = resources.getString(R.string.n_dBm, rssi)
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {}
            override fun onStopTrackingTouch(seekBar: SeekBar) {}
        })
        btn_search.setOnClickListener { submit(true) }

        iv_beacon_arrow.setOnClickListener { //If after clicking arrow is up then...
            if (iv_beacon_arrow.tag == "ARROW_UP") {
                iv_beacon_arrow.setImageDrawable(ContextCompat.getDrawable(requireContext(), R.drawable.ic_arrow_down_on))
                rv_beacon_type.visibility = View.GONE
                iv_beacon_arrow.tag = "ARROW_DOWN"
            } else {
                iv_beacon_arrow.setImageDrawable(ContextCompat.getDrawable(requireContext(), R.drawable.ic_arrow_up_on))
                rv_beacon_type.visibility = View.VISIBLE
                iv_beacon_arrow.tag = "ARROW_UP"
            }
        }

        iv_save_arrow.setOnClickListener {
            if (iv_save_arrow.tag == "ARROW_UP") {
                iv_save_arrow.setImageDrawable(ContextCompat.getDrawable(requireContext(), R.drawable.ic_arrow_down_on))
                rv_saved_searches.visibility = View.GONE
                iv_save_arrow.tag = "ARROW_DOWN"
            } else {
                iv_save_arrow.setImageDrawable(ContextCompat.getDrawable(requireContext(), R.drawable.ic_arrow_up_on))
                rv_saved_searches.visibility = View.VISIBLE
                iv_save_arrow.tag = "ARROW_UP"
            }
        }
    }

    private fun prepareFilterName(): String {
        return StringBuilder().apply {
            val separator = " + "

            if (rssiFlag) append("RSSI > ").append(tv_rssi_value.text).append(separator)

            val selectedBeacons = selectedBeacons
            if (selectedBeacons.isNotEmpty()) {
                append("Beacon type ")
                for (i in selectedBeacons.indices) {
                    append("\"").append(resources.getString(selectedBeacons[i].nameResId)).append("\"")
                    if (i != selectedBeacons.size - 1) append(", ")
                }
                append(separator)
            }

            if (et_search_device_name.text.toString() != "") append("Name \"").append(et_search_device_name.text.toString()).append("\"").append(separator)
            if (et_search_packet_content.text.toString() != "") append("Packet content \"").append(et_search_packet_content.text.toString()).append("\"").append(separator)
            if (cb_only_favourites.isChecked) append("Only favourites").append(separator)
            if (cb_only_connectable.isChecked) append("Only connectable").append(separator)
            if (cb_only_bonded.isChecked) append("Only bonded")
            if (toString().endsWith(separator)) return substring(0, length - 3)

        }.toString()
    }

    private fun getRssiValue(progress: Int): Int {
        return progress - 100
    }

    private fun resetUi() {
        et_search_device_name.setText("")
        et_search_packet_content.setText("")
        sb_rssi.progress = 0
        tv_rssi_value.setText(R.string.filter_rssi_not_set)
        rssiFlag = false
        cb_only_favourites.isChecked = false
        cb_only_connectable.isChecked = false
        cb_only_bonded.isChecked = false
        beaconTypeadapter.selectBeacons(emptyList())
    }

    private fun fillFields(filterDeviceParams: FilterDeviceParams) {
        et_search_device_name.setText(filterDeviceParams.name)
        et_search_packet_content.setText(filterDeviceParams.advertising)
        sb_rssi.progress = filterDeviceParams.rssiValue + 100
        sb_rssi.refreshDrawableState()
        val rssi = filterDeviceParams.rssiValue
        rssiFlag = filterDeviceParams.isRssiFlag
        tv_rssi_value.text = if (rssiFlag) resources.getString(R.string.n_dBm, rssi) else context?.getString(R.string.filter_rssi_not_set)
        val selectedBleFormats = filterDeviceParams.bleFormats
        beaconTypeadapter.selectBeacons(selectedBleFormats)
        cb_only_favourites.isChecked = filterDeviceParams.isOnlyFavourite
        cb_only_connectable.isChecked = filterDeviceParams.isOnlyConnectable
        cb_only_bonded.isChecked = filterDeviceParams.isOnlyBonded
    }

    private val selectedBeacons: List<BleFormat>
        get() {
            val selectedBeacons: MutableList<BleFormat> = ArrayList()
            for (b in beaconTypeadapter.beaconTypeList) {
                if (b.isChecked) selectedBeacons.add(b.bleFormat)
            }
            return selectedBeacons
        }

    private fun prepareFilterDeviceParam(): FilterDeviceParams {
        val name = et_search_device_name.text.toString()
        val advertising = et_search_packet_content.text.toString()
        val rssi = getRssiValue(sb_rssi.progress)
        return FilterDeviceParams(prepareFilterName(), name, advertising, rssi, rssiFlag, selectedBeacons, cb_only_favourites.isChecked, cb_only_connectable.isChecked, cb_only_bonded.isChecked)
    }

    private fun submit(closeToolbar: Boolean) {
        val filterDeviceParams = prepareFilterDeviceParam()
        sharedPrefUtils.lastFilter = filterDeviceParams
        toolbarCallback?.submit(filterDeviceParams, closeToolbar)
    }
}